### PR TITLE
vstart_runner: refactor and reuse code from qa/cephfs

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Reduces vstart_runner.py's size by 200 lines so far.

- [x] Refactor FuseMount, KernelMount and CephFSMount
- [x] Refactor LocalFilesystem
- [x] Reuse FuseMount's method in LocalFuseMount
- [x] Reuse KernelMount's method in LocalKernelMount
- [x] Move common methods
	from LocalFuseMount and LocalKernelMount to LocalCephFSMount
	from FuseMount and KernelMount to CephFSMount

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by batrick** [comment]
@sidharthanup please review this.

**Comment by rishabh-d-dave** [comment]
@batrick @sidharthanup ping

**Comment by sidharthanup** [comment]
@rishabh-d-dave Please do a rebase.

**Comment by rishabh-d-dave** [comment]
Rebased again.

**Comment by rishabh-d-dave** [comment]
CI failures look unrelated, re-triggering...

**Comment by rishabh-d-dave** [comment]
jenkins test docs

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
jenkins test docs

**Comment by rishabh-d-dave** [comment]
jenkins test api

File: qa/tasks/cephfs/fuse_mount.py [hunk 1]
```diff
@@ -173,17 +194,31 @@ def list_connections():
...
-        # forcing an unmount
+        self._record_our_fuse_conn(pre_mount_conns, post_mount_conns)
+
+    @property
+    def _fuse_conn_check_timeout(self):
+        mount_wait = self.client_config.get('mount_wait', 0)
+        if mount_wait > 0:
+            log.info("Fuse mount waits {0} seconds before checking /sys/".format(mount_wait))
+            time.sleep(mount_wait)
+        timeout = int(self.client_config.get('mount_timeout', 30))
```
Comment (lxbsz) [inline qa/tasks/cephfs/fuse_mount.py:199]: There maybe no need to get it from the config everytime, save it in `self.timeout` or something else. If the `self.timeout` is not None just return it ?

File: qa/tasks/cephfs/fuse_mount.py [hunk 2]
```diff
@@ -58,106 +65,120 @@ def _mount(self, mntopts, check_status):
...
+            mount_cmd += ["--client_fs=" + self.cephfs_name]
         if mntopts:
-            fuse_cmd += mntopts
-        fuse_cmd.append(self.hostfs_mntpt)
+            mount_cmd += mntopts
 
-        cwd = self.test_dir
+        return mount_cmd
+
+    def _get_nsenter_args(self):
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply qa/tasks/cephfs/fuse_mount.py]: Okay I saw you have made it a generic method in the following patch.

File: qa/tasks/cephfs/kernel_mount.py [hunk 3]
```diff
@@ -41,33 +43,43 @@ def mount(self, mntopts=[], createfs=True, check_status=True, **kwargs):
...
+            log.info('mount command failed')
+            if check_status:
+                raise
+            else:
+                return (e, mountcmd_stdout.getvalue(),
+                        mountcmd_stderr.getvalue())
+        log.info('mount command passed')
 
-    def _run_mount_cmd(self, mntopts, check_status):
+    def _get_mount_cmd(self, mntopts):
```
[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/kernel_mount.py:73]: That would be better but passing argument would be strange to read -

```
    @property
    def mount_cmd(self):
        return self._mount_cmd

    @mount_cmd.setter
    def mount_cmd(self, mntopts):

    def _run_mount_cmd(self, mntopts, check_status):
        self.mount_cmd = mntopts # the "strange" line
        mountcmd_stdout, mountcmd_stderr = StringIO(), StringIO()
```

`mount_cmd = self._get_mount_cmd(mntopts)` reads better IMHO.

File: qa/tasks/cephfs/kernel_mount.py [hunk 4]
```diff
@@ -84,27 +96,27 @@ def _run_mount_cmd(self, mntopts, check_status):
...
-                  '--net=/var/run/netns/{0}'.format(self.netns_name)]
-        cmdargs = prefix + ['/bin/mount', '-t', 'ceph', mount_dev,
-                            self.hostfs_mntpt, '-v', '-o', opts]
+        mount_cmd += self._mount_bin + [mount_dev, self.hostfs_mntpt, '-v',
+                                        '-o', opts]
 
-        mountcmd_stdout, mountcmd_stderr = StringIO(), StringIO()
+        return mount_cmd
+
+    def _get_nsenter_args(self):
```
[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/kernel_mount.py]: Done.

File: qa/tasks/cephfs/fuse_mount.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/fuse_mount.py]: Done.

File: qa/tasks/cephfs/fuse_mount.py (hunk 1 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/fuse_mount.py:200]: Moved. Thanks!

File: qa/tasks/cephfs/kernel_mount.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lxbsz) [reply qa/tasks/cephfs/kernel_mount.py:73]: That's okay for me.

**Review by lxbsz — APPROVED** [review]
LGTM.

**Comment by batrick** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
http://pulpito.front.sepia.ceph.com/rishabh-2020-12-03_18:20:21-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi

**Comment by rishabh-d-dave** [comment]
@batrick 
Tests with teuthology passed. Patch for fuse_mount.py and mount.py have been tested. What suite should I run to test kernel_mount,py patch?

**Comment by batrick** [comment]
I will include it in my next run this weekend. I will review this today.

**Comment by rishabh-d-dave** [comment]
Great, thanks!

**Comment by rishabh-d-dave** [comment]
@batrick There's no need for a tracker ticket for this, right?

**Comment by batrick** [comment]
> @batrick There's no need for a tracker ticket for this, right?

If it won't be backported, no.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
@rishabh-d-dave commit organization and code changes look good to me. Thanks.

**Comment by rishabh-d-dave** [comment]
@batrick Yay, thanks!

**Comment by batrick** [comment]
waiting on: https://pulpito.ceph.com/pdonnell-2020-12-08_00:31:36-fs-wip-pdonnell-testing-20201207.222120-distro-basic-smithi/

**Review by batrick — CHANGES_REQUESTED** [review]
Something broke the fuse client setup with valgrind:

https://pulpito.ceph.com/pdonnell-2020-12-08_00:31:36-fs-wip-pdonnell-testing-20201207.222120-distro-basic-smithi/5691666/

and others. I don't see what's immediately wrong. Suggest you look at past successful runs and compare.

**Comment by rishabh-d-dave** [comment]
@batrick Cool

**Comment by rishabh-d-dave** [comment]
The following two jobs exercise all changes to `qa/tasks/cephfs/fuse_mount.py`, `qa/tasks/cephfs/kernel_mount.py` and `qa/tasks/cephfs/mount.py` -

https://pulpito.ceph.com/rishabh-2021-02-11_13:58:57-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
https://pulpito.ceph.com/rishabh-2021-02-11_13:27:21-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/

And I ran test_cephfs_shell.TestCD.test_cd_with_args and test_acls using vstart_runner.py to exercise modifications in vstart_runner.py.

**Comment by rishabh-d-dave** [comment]
> The following two teuthology jobs exercise all changes to `qa/tasks/cephfs/fuse_mount.py`, `qa/tasks/cephfs/kernel_mount.py` and `qa/tasks/cephfs/mount.py` -
> 
> https://pulpito.ceph.com/rishabh-2021-02-11_13:58:57-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
> https://pulpito.ceph.com/rishabh-2021-02-11_13:27:21-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
> 
> And I ran test_cephfs_shell.TestCD.test_cd_with_args and test_acls using vstart_runner.py to exercise modifications in vstart_runner.py. So, I think this PR is ready for merge.

@full

**Comment by rishabh-d-dave** [comment]
> The following two teuthology jobs exercise all changes to `qa/tasks/cephfs/fuse_mount.py`, `qa/tasks/cephfs/kernel_mount.py` and `qa/tasks/cephfs/mount.py` -
> 
> https://pulpito.ceph.com/rishabh-2021-02-11_13:58:57-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
> https://pulpito.ceph.com/rishabh-2021-02-11_13:27:21-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
> 
> And I ran test_cephfs_shell.TestCD.test_cd_with_args and test_acls using vstart_runner.py to exercise modifications in vstart_runner.py. So, I think this PR is ready for merge.

@fullerdj

**Comment by rishabh-d-dave** [comment]
@batrick I tested this PR, here are the results - https://github.com/ceph/ceph/pull/37711#issuecomment-777999657. Can you please review/merge this PR? Thanks!

**Review by batrick — CHANGES_REQUESTED** [review]
Several things:

- Move `73c92f16cd3466f6b325e32ad01cd97c373f376e` to a separate PR.
- Drop `vstart_runner: move few methods definitions`. This is code churn that confuses `git blame` needlessly.
- `vstart_runner: remove duplicate methods from LocalCephManager` should be in a separate PR.
- `qa/cephfs: refactor class LocalFilesystem` moves code lines needlessly, more code churn.
- `qa/cephfs: refactor kernel_mount.py` is too vague about what it's doing. Please update the commit explaining what the goal is. Same for `qa/cephfs: refactor fuse_mount.py`.

**Comment by rishabh-d-dave** [comment]
@batrick Thanks for the review.

> Several things:
> 
>     * Move `73c92f16cd3466f6b325e32ad01cd97c373f376e` to a separate PR.

https://github.com/ceph/ceph/pull/39681

>     * Drop `vstart_runner: move few methods definitions`. This is code churn that confuses `git blame` needlessly.

Dropped.
 
>     * `vstart_runner: remove duplicate methods from LocalCephManager` should be in a separate PR.
> 

https://github.com/ceph/ceph/pull/39682

>     * `qa/cephfs: refactor class LocalFilesystem` moves code lines needlessly, more code churn.

Dropped.

>     * `qa/cephfs: refactor kernel_mount.py` is too vague about what it's doing. Please update the commit explaining what the goal is. Same for `qa/cephfs: refactor fuse_mount.py`.

Rewrote the commit message to make it more specific.

**Comment by rishabh-d-dave** [comment]
@batrick 

Gave tests a run with teuthology for both FUSE and kernel client; both ran successfully -

http://pulpito.front.sepia.ceph.com/rishabh-2021-02-25_12:24:33-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/

**Review by batrick — CHANGES_REQUESTED** [review]
This looks good now. Just one oddity I noticed: commit `vstart_runner: reuse KernelMount methods in LocalKernelMount` introduced a `self.Filesystem` that was later deleted. It didn't look necessary, is that right? If so, can you remove it from the git history?

**Comment by rishabh-d-dave** [comment]
Ran tests once again just to be sure - http://pulpito.front.sepia.ceph.com/rishabh-2021-03-01_10:36:20-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/.

File: qa/tasks/vstart_runner.py [hunk 5]
```diff
@@ -745,164 +681,85 @@ def _run_python(self, pyscript, py_version='python'):
...
-    def _prefix(self):
-        return BIN_PREFIX
 
-    def _asok_path(self):
-        # In teuthology, the asok is named after the PID of the ceph-fuse process, because it's
-        # run foreground.  When running it daemonized however, the asok is named after
-        # the PID of the launching process, not the long running ceph-fuse process.  Therefore
-        # we need to give an exact path here as the logic for checking /proc/ for which
-        # asok is alive does not work.
+class LocalKernelMount(LocalCephFSMount, KernelMount):
```
Comment (lxbsz) [inline qa/tasks/vstart_runner.py:705]: LocalKernelMount is redefined twice.

Comment (rishabh-d-dave) [reply qa/tasks/vstart_runner.py:681]: FIxed, thanks!

**Comment by rishabh-d-dave** [comment]
@batrick ping

Here's our last conversation on this PR - https://github.com/ceph/ceph/pull/37711#issuecomment-785963532

**Comment by rishabh-d-dave** [comment]
@batrick 

Tested again, ran successfully -
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-08_14:48:25-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-08_14:49:08-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/

**Review by batrick — APPROVED** [review]

**Pull request merged**